File: DSA_Level-1/Basics Of Programming/2D Arrays/Rotate By 90 Degree/RotateBy90Degree_Solution.java
import java.io.*;
import java.util.*;

public class Main {

  public static void main(String[] args) throws Exception {
    Scanner scn = new Scanner(System.in);
    int n = scn.nextInt();
    int[][] nums = new int[n][n];

    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < n; ++j) {
        nums[i][j] = scn.nextInt();
      }
    }
    
    rotate90(nums);
    display(nums);
  }
  
  public static void rotate90(int[][] nums){
      //Transpose

    for (int i = 0; i < nums.length; ++i) {
      for (int j = i; j < nums[i].length; ++j) {
        int temp = nums[i][j];
        nums[i][j] = nums[j][i];
        nums[j][i] = temp;
      }
    }


    //Reverse

    for (int i = 0; i < nums.length; ++i) {
      int left = 0, right = nums[i].length - 1;
      while (left < right) {
        int temp = nums[i][left];
        nums[i][left] = nums[i][right];
        nums[i][right] = temp;
        ++left;
        --right;
      }
    }
  }

  public static void display(int[][] nums) {
    for (int i = 0; i < nums.length; i++) {
      for (int j = 0; j < nums[0].length; j++) {
        System.out.print(nums[i][j] + " ");
      }
      System.out.println();
    }
  }

}

File: DSA_Level-1/Basics Of Programming/2D Arrays/Rotate By 90 Degree/RotateBy90Degree_Solution.js

function rotateBy90Deg(nums){
    //Transpose

    for(let i=0;i<nums.length;++i){
        for(let j=i;j<nums[i].length;++j){
            let temp=nums[i][j];
            nums[i][j]=nums[j][i];
            nums[j][i]=temp;
        }
    }


    //Reverse

    for(let i=0;i<nums.length;++i){
        let left=0,right=nums[i].length-1;
        while(left<right){
            let temp=nums[i][left];
            nums[i][left]=nums[i][right];
            nums[i][right]=temp;
            ++left;
            --right;
        }
    }    

    console.log(nums);
}

rotateBy90Deg([[1,2,3],[4,5,6],[7,8,9]]);